make rust-analyzer use a dedicated build directory

tshepang opened the pull request:
inspired by https://github.com/rust-lang/rust/pull/132794

jieyouxu commented:
I feel like this was discussed [#general > ✔ Setup multiple editors for work on standard lib? @ 💬](https://rust-lang.zulipchat.com/#narrow/channel/122651-general/topic/.E2.9C.94.20Setup.20multiple.20editors.20for.20work.20on.20standard.20lib.3F/near/479096525) but I don't remember why we used a top-level `build-rust-analyzer/` as opposed to nesting inside of `build/`... @Zalathar do you happen to remember why? (It's been a while)

Zalathar commented:
It has unintuitive interactions with `x clean` and caching, for example. IIRC `x clean` won’t delete your download cache, but it will delete your nested RA download cache, because it’s just indiscriminately deleting a subdirectory it doesn’t know about.

If people choose to inflict those risks on themselves, fine, but I don’t think a nested RA build directory is a responsible default.

jieyouxu commented:
Oh right. I think that's the caveat I was remembering...

tshepang commented:
would it be good to get `x clean` not delete what it does not recognize

mati865 commented:
Or even just avoid deleting `build/rust-analyzer`.

Zalathar commented:
> would it be good to get `x clean` not delete what it does not recognize

Bootstrap “owns” the build directory, so I think it's reasonable for `x clean` to delete anything it doesn't know about, on the assumption that it's junk left over from an earlier version of bootstrap.

> Or even just avoid deleting `build/rust-analyzer`.

IIRC, one of the motivations people commonly cite for using `build/rust-analyzer` is that they *want* `x clean` to also clean their RA build directory. (And it does accomplish that, as a consequence of deleting `build/rust-analyzer` entirely.)

Zalathar commented:
Anyhow, I strongly agree that our default IDE configs really should be using a separate build dir (e.g. `build-rust-analyzer`).

tshepang commented:
I recently found that `x clean` does not in fact remove build/rust-analyzer

Mark-Simulacrum commented:
Could you file a compiler [MCP](https://github.com/rust-lang/compiler-team/issues/new/choose) for this? I'd like to see some more eyes on this before we move in this direction.

lolbinarycat commented:
Ah, right, this will still improve things for other editors (this PR was linked to me as "changing the default for emacs"), so there's no point blocking it on emacs being fixed.

tshepang commented:
made the move to build-rust-analyzer more comprehensive... it did not cover rustfmt and procmacro settings

should be ready now @Mark-Simulacrum

bors commented:
:pushpin: Commit fdd2352d2c7d3a8c339f28732a79d0c976731fa1 has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum fdd2352d2c7d3a8c339f28732a79d0c976731fa1 -->
<!-- homu: {"type":"Approved","sha":"fdd2352d2c7d3a8c339f28732a79d0c976731fa1","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #143888) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

tshepang commented:
needs another approval...

tshepang committed:
make rust-analyzer settings use dedicated directory

This avoids rust-analyzer having to wait for a build lock due to ./x
running other commands (and the other way around).

fee1-dead commented:
Please use `[@]rustbot ready` to indicate that your PR is ready for review next time

@bors r=Mark-Simulacrum,fee1-dead

bors commented:
:pushpin: Commit 24144ff499852b0fab63cf0c93d4c8d8b2fad8ef has been approved by `Mark-Simulacrum,fee1-dead`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum,fee1-dead 24144ff499852b0fab63cf0c93d4c8d8b2fad8ef -->
<!-- homu: {"type":"Approved","sha":"24144ff499852b0fab63cf0c93d4c8d8b2fad8ef","approver":"Mark-Simulacrum,fee1-dead","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Zalathar commented:
Bors hasn't noticed that this was merged.

@bors r-
